File: session09/Matrix.java
public class Matrix {
	private double[] m;
	private int rows, cols;

	public Matrix(int r, int c) {
		m = new double[r*c];
		rows = r; cols = c;
	}

	/*
		0 5 0 0
		0 0 0 0
		0 0 0 0



	 */
	public Matrix add(Matrix b) {
		if (b.rows != this.rows || b.cols != this.cols)
			throw new RuntimeException("Bad Size");
		Matrix ans = new Matrix(this.rows, this.cols);
		for (int i = 0; i < rows; i++)
			for (int j = 0; j < cols; j++)
				ans.m[i*cols+j] = this.m[i*cols+j] + b.m[i*cols+j];
		
    for (int i = 0; i < rows * cols; i++)
			ans.m[i] = m[i] + b.m[i];
	}

	public void set(int i, int j, double v) {
		m[i*cols+j] = v;
	}
	public double get(int i, int j) {
		return m[i*cols+j];
	}
	public static void main(String[] args) {
		Matrix m1 = new Matrix(3,4);
		m1.set(0,1, 5.0);
		Matrix m2 = new Matrix(3,4);
		m2.set(2,0, 2.5);
		Matrix m3 = m1.add(m2); // m1 + m2
		Matrix m4 = Matrix.add(m1,m2);

	}
}

File: session09/Matrix.cc
class Matrix {
private:
	int rows, cols;
	double* m;
	// this one doesn't initialize, you had better know what you are doing!!!
	Matrix(int r, int c, char*) : rows(r), cols(c), m(new double[r*c]) {
	}
public:
  Matrix(int r, int c, double val = 0) : rows(r), cols(c),
																				 m(new double[rows*cols]) {
		for (int i = 0; i < rows*cols; ++i)
			m[i] = val;
	}

	~Matrix() {
		delete [] m;
	}
	Matrix(const Matrix& orig);
	Matrix& operator =(const Matrix& orig);
  Matrix(Matrix&& orig) : rows(orig.rows), cols(orig.cols), m(orig.m) {
		orig.m = nullptr;
	}
	
	friend Matrix operator + (const Matrix& a, const Matrix& b) {
		if (a.rows != b.rows || a.cols != b.cols)
			throw "Bad size";
    Matrix ans(a.rows, a.cols, "internal");
#if 0
		// slow way, thinking like a rectangle
		for (int i = 0; i < a.rows; ++i)
			for (int j = 0; j < a.cols; ++j)
				ans.m[i*cols+j] = a.m[i*cols+j] + b.m[i*cols+j];
#endif
		//faster (just a bit) linearizing
		for (int i = 0; i < a.rows*a.cols; ++i)
			ans.m[i] = a.m[i] + b.m[i];

		return ans;
	}

};

int main() {
	Matrix m1(3,4, 1);
	Matrix m2(3,4, 2);
	Matrix m3 = m1 + m2;
										
